File: Easy/Count pairs with given sum/count-pairs-with-given-sum.cpp
//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function template for C++

class Solution{   
public:
    int getPairsCount(int arr[], int n, int k) {
        // code here
    //     sort(arr,arr+n);
    //     int count =0;
    //     int l=0;
    //     int r=n-1;
    //     while(l!=r){
    //         if(arr[l]+arr[r]==k){
    //             count++;
    //             // l++;
    //             r--;
    //         }
    //         else if(arr[l]+arr[k]>k)
    //             r--;
    //         else if(arr[l]+arr[k]<k)
    //             l++;
    // }
    // return count;
     unordered_map<int, int> mp;
            
            int count = 0;
            for (int i = 0; i < n; i++)
            {
                int diff = k - arr[i];
                if (mp[diff] > 0)
                    count += mp[diff];
                mp[arr[i]]++;
            }
            return count;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, k;
        cin >> n >> k;
        int arr[n];
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.getPairsCount(arr, n, k);
        cout << ans << "\n";
    }
    
    return 0;
}
// } Driver Code Ends

File: Easy/Count pairs with given sum/count-pairs-with-given-sum.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let [n,k] = readLine().trim().split(" ").map((x) => parseInt(x)); 
        let arr = readLine().trim().split(" ").map((x) => parseInt(x)); 
        let obj = new Solution();
        let res = obj.getPairsCount(arr,n,k);
        console.log(res);
    }
}
// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number[]} arr
 * @param {number} n
 * @param {number} k
 * @return {number}
*/

class Solution {
    getPairsCount(arr,n,k){
       const mp = {}; // Using an object as a map
    let count = 0;

    for (let i = 0; i < arr.length; i++) {
        const diff = k - arr[i];
        if (mp[diff] > 0) {
            count += mp[diff];
        }
        mp[arr[i]] = (mp[arr[i]] || 0) + 1;
    }

    return count;
    }
}